### PR TITLE
Improve the accuracy of waits on Chibios

### PR DESCRIPTION
**Pull request by skullydazed**
## Description

On Chibios `wait_us()` is currently not very accurate for small values. This changes to the method @zvecr identified.

This needs more testing before we roll it out.

## Types of Changes

- [x] Core
- [x] Bugfix

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: tmk_core/common/wait.h [hunk 1]
```diff
@@ -13,22 +13,9 @@ extern "C" {
...
-#    define wait_us(us)                     \
-        do {                                \
-            if (us != 0) {                  \
-                chThdSleepMicroseconds(us); \
-            } else {                        \
-                chThdSleepMicroseconds(1);  \
-            }                               \
-        } while (0)
+// FIXME: This doesn't compile if `NKRO_ENABLED = no`
+#    define wait_ms(ms) chSysPolledDelayX(MS2RTC(STM32_SYSCLK, x))
```
Comment (drashna) [inline tmk_core/common/wait.h]:
```suggestion
#    define wait_ms(ms) chSysPolledDelayX(MS2RTC(STM32_SYSCLK, ms))
```

File: tmk_core/common/wait.h [hunk 2]
```diff
@@ -13,22 +13,9 @@ extern "C" {
...
+#    define wait_us(us) chSysPolledDelayX(US2RTC(STM32_SYSCLK, x))
```
Comment (drashna) [inline tmk_core/common/wait.h]:
```suggestion
#    define wait_us(us) chSysPolledDelayX(US2RTC(STM32_SYSCLK, us))
```

**Comment by skullydazed** [comment]
Yup, the dangers of redoing PR's.

**Comment by drashna** [comment]
Unfortunately, it looks like this is very dependant on the chibiOS config, which means changing all of the boards and hoping it works.

**Comment by zvecr** [comment]
@drashna got any specifics of scenarios this wont work for? And what parts of config its dependent on?

File: tmk_core/common/wait.h [hunk 3]
```diff
@@ -13,22 +13,9 @@ extern "C" {
 #    define wait_us(us) _delay_us(us)
 #elif defined PROTOCOL_CHIBIOS
 #    include "ch.h"
```
Comment (zvecr) [inline tmk_core/common/wait.h]: Had to add the following to get it to work with my serial PR.
```suggestion
#    include "ch.h"
#    include "hal.h"
```

**Comment by drashna** [comment]
@zvecr Just travis.

**Comment by zvecr** [comment]
@drashna not sure what you mean about travis, i was referring to:
>  it looks like this is very dependant on the chibiOS config, which means changing all of the boards and hoping it works

Though i havent vastly tested systick vs tickless right now, i have gone f303 with tickless and f103 with systick and both give the expected result for `wait_us` timings.

**Comment by drashna** [comment]
Well, right now, Travis is erroring out on every F072 chip, and every Kinesis chip (Teensy ARM).

**Comment by zvecr** [comment]
Yeah teensy looks like its going to be a bit more of a pain, and f072 is something to do with `PORT_SUPPORTS_RT`. I have some test hardware so will have a tinker.

**Comment by drashna** [comment]
Checking the makefile stuff, there isn't really anything that we could use to ID the chips right now. 

However, we could add something to this block: 
https://github.com/qmk/qmk_firmware/blob/68cf2725aa30eaa18db9c6149e5e361860db01d7/tmk_core/common.mk#L32-L57

That would at least give us a decent way to identify the MCU, and base the handling on that.

**Comment by skullydazed** [comment]
We're setting `MCU = STM32F303` in `rules.mk`. Could we use that to identify the MCU?

**Comment by drashna** [comment]
In makefiles, yes.  But I don't think we're setting a define for it.  At least not that I could find

**Comment by zvecr** [comment]
We do the following in the ARM I2C driver
https://github.com/qmk/qmk_firmware/blob/e8453bbc1218a2cb40d4eac6cb543bd0c45d3de9/drivers/arm/i2c_master.h#L30-L32

**Comment by drashna** [comment]
@zvecr We have that, but I don't see where those are actually defined.   I know whe have: `EEPROM_EMU_STM32F303xC` and the like, but not ... those. 

We'd need to add something like:
```
TMK_COMMON_DEFS += -D$(MCU_FAMILY)
```
(assuming this even works)

**Comment by zvecr** [comment]
The i2c stuff works fine, I've tested it. it's using the chibios board.h defines.

**Comment by drashna** [comment]
Aah, okay, it's defined a lot deeper into the chibiOS submodule, which is why I wasn't finding it. 

And yeah, it is defined by the HAL files in chibiOS.   So should definitely be fine to use. 


Specifically in:
```
lib/chibios/os/hal/ports/STM32/(mcu)/hal_lld.h
```

File: tmk_core/common/wait.h [hunk 4]
```diff
@@ -13,22 +13,10 @@ extern "C" {
...
-            if (us != 0) {                  \
-                chThdSleepMicroseconds(us); \
-            } else {                        \
-                chThdSleepMicroseconds(1);  \
-            }                               \
-        } while (0)
+#    include "hal.h"
+// FIXME: This doesn't compile if `NKRO_ENABLED = no`
+#    define wait_ms(ms) chSysPolledDelayX(MS2RTC(STM32_SYSCLK, ms))
+#    define wait_us(us) chSysPolledDelayX(US2RTC(STM32_SYSCLK, us))
```
Comment (tzarc) [inline tmk_core/common/wait.h:19]:
```suggestion
#    define wait_ms(ms) chSysPolledDelayX(TIME_MS2I(ms))
#    define wait_us(us) chSysPolledDelayX(TIME_US2I(us))
```

**Review by tzarc — CHANGES_REQUESTED** [review]

**Comment by tzarc** [comment]
Removed "on hold" so stalebot can eventually close this.

**Pull request closed without merging**